File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/mainmenus/marketplace/views/ThemeSingleCardView.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities.mainmenus.marketplace.views;

import android.content.Context;
import android.content.res.Resources;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatTextView;

import com.TritiumGaming.phasmophobiaevidencepicker.R;
import com.TritiumGaming.phasmophobiaevidencepicker.data.utilities.ColorUtils;
import com.TritiumGaming.phasmophobiaevidencepicker.firebase.firestore.objects.theme.theme.MarketSingleTheme;
import com.google.android.material.card.MaterialCardView;

public class ThemeSingleCardView extends MaterialCardView {

    @Nullable
    private MarketSingleTheme theme = null;

    public ThemeSingleCardView(@NonNull Context context) {
        super(context, null);
    }

    public ThemeSingleCardView(@NonNull Context context, AttributeSet attrs) {
        super(context, attrs);

        init(context);
    }

    public ThemeSingleCardView(@NonNull Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        init(context);
    }

    public void init(@Nullable Context context) {
        inflate(context, R.layout.item_marketplace_theme, this);

        setLayoutParams(
                new LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT));

        int strokeColor =
                ColorUtils.getColorFromAttribute(getContext(), R.attr.backgroundColorOnBackground);

        setRadius(16);

        setStrokeColor(ColorUtils.interpolate(
                getResources().getColor(R.color.white),
                strokeColor,
                .25f));
        setStrokeWidth(3);

        setUseCompatPadding(true);
        setClipToPadding(false);
    }

    public void setPurchasable() {
        if(theme == null) {
            return;
        }

        if(theme.isUnlocked())
            setVisibility(GONE); {
        }

    }

    public void setCreditCost() {
        AppCompatTextView label_credits = findViewById(R.id.label_credits_cost);

        if(theme == null || label_credits == null) {
            return;
        }

        label_credits.setText(String.valueOf(theme.getBuyCredits()));
    }

    public long getCreditCost() {
        return theme.getBuyCredits();
    }

    public void setTheme(MarketSingleTheme theme) {
        this.theme = theme;

        setPurchasable();
        setCreditCost();
    }

    public void validate() {
        setPurchasable();
    }

    public void setBuyButtonListener(OnClickListener buyButtonListener) {
        View buyButton = findViewById(R.id.button_transactItem);

        if(buyButton != null) {
            buyButton.setOnClickListener(buyButtonListener);
        }
    }
}

File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/mainmenus/marketplace/views/items/ThemeSingleCardView.kt
package com.TritiumGaming.phasmophobiaevidencepicker.activities.mainmenus.marketplace.views.items

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.appcompat.widget.AppCompatTextView
import com.TritiumGaming.phasmophobiaevidencepicker.R
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.models.firestore.theme.theme.MarketSingleThemeModel
import com.TritiumGaming.phasmophobiaevidencepicker.utils.ColorUtils.getColorFromAttribute
import com.TritiumGaming.phasmophobiaevidencepicker.utils.ColorUtils.interpolate
import com.google.android.material.card.MaterialCardView

class ThemeSingleCardView : MaterialCardView {
    private var theme: MarketSingleThemeModel? = null

    constructor(context: Context) : super(context, null)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) { initView(context) }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
            super(context, attrs, defStyleAttr) { initView(context) }

    fun initView(context: Context?) {
        inflate(context, R.layout.item_marketplace_theme, this)

        layoutParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT)

        val strokeColor =
            getColorFromAttribute(getContext(), R.attr.backgroundColorOnBackground)

        radius = 16f

        setStrokeColor(interpolate(resources.getColor(R.color.white), strokeColor, .25f))
        strokeWidth = 3

        useCompatPadding = true
        clipToPadding = false
    }

    private fun setPurchasable() {
        theme?.let { theme ->
            if (theme.isUnlocked) visibility = GONE
            run {}
        }
    }

    fun setCreditCost() {
        val creditsLabel = findViewById<AppCompatTextView>(R.id.label_credits_cost)

        theme?.let { theme -> creditsLabel?.text = theme.buyCredits.toString() }
    }

    val creditCost: Long
        get() = theme?.buyCredits ?: 0

    fun setTheme(theme: MarketSingleThemeModel?) {
        this.theme = theme

        setPurchasable()
        setCreditCost()
    }

    fun validate() { setPurchasable() }

    fun setBuyButtonListener(buyButtonListener: OnClickListener?) {
        val buyButton = findViewById<View>(R.id.button_transactItem)

        buyButton?.setOnClickListener(buyButtonListener)
    }
}
